lib/spack/llnl/util/tty/pty.py

```diff
@@ -0,0 +1,287 @@
+
+    def tstp(self):
+        """Send SIGTSTP to the controlled process."""
+        self.hline("tstp")
+        os.killpg(self.pgid, signal.SIGTSTP)
+
+    def cont(self):
+        self.hline("cont")
+        os.killpg(self.pgid, signal.SIGCONT)
+
+    def fg(self):
+        self.hline("fg")
+        os.tcsetpgrp(self.master_fd, os.getpgid(self.pid))
+
+    def bg(self):
+        self.hline("bg")
+        with log.ignore_signal(signal.SIGTTOU):
+            os.tcsetpgrp(self.master_fd, os.getpgrp())
+
+    def write(self, byte_string):
+        self.hline("write '%s'" % byte_string.decode("utf-8"))
+        os.write(self.master_fd, byte_string)
+
+    def wait(self, condition):
+        start = time.time()
+        while (((time.time() - start) < self.timeout) and not condition()):
+            time.sleep(1e-2)
+        assert condition()
+
+    def wait_enabled(self):
+        self.wait(lambda: self.input_on() and not self.background())
+
+    def wait_disabled(self):
+        self.wait(lambda: not self.input_on() and self.background())
+
+    def wait_disabled_fg(self):
+        self.wait(lambda: not self.input_on() and not self.background())
+
+    def proc_status(self):
+        status = self.ps("-p", str(self.pid), "-o", "stat", output=str)
+        status = re.split(r"\s+", status.strip(), re.M)
+        return status[1]
+
+    def wait_stopped(self):
+        self.wait(lambda: "T" in self.proc_status())
+
+    def wait_running(self):
+        status = self.proc_status()
+        self.wait(lambda: "T" not in status)
+
+
+def _child_process(
+        tty_name, stdout_fd, stderr_fd, ready, child_function, attrs):
+    """Child process wrapper for PseudoShell.
+
+    Handles the mechanics of setting up a PTY, then calls
+    ``child_function``.
+
+    """
+    # new process group, like a command or pipeline launched by a shell
+    os.setpgrp()
+
+    # take controlling terminal and set up pty IO
+    stdin_fd = os.open(tty_name, os.O_RDWR)
+    os.dup2(stdin_fd, sys.stdin.fileno())
+    os.dup2(stdout_fd, sys.stdout.fileno())
+    os.dup2(stderr_fd, sys.stderr.fileno())
+    os.close(stdin_fd)
+
+    if attrs.get("debug"):
+        sys.stderr.write("child: stdin.isatty(): %s\n" % sys.stdin.isatty())
+
+    # tell the parent that we're really running
+    if attrs.get("debug"):
+        sys.stderr.write("child: ready!\n")
+    ready.value = True
+
+    child_function(attrs)
+
+
+def _master_process(master_function, child_function, attrs):
+    """Master process wrapper for PseudoShell.
+
+    Handles the mechanics of setting up a PTY, then calls
+    ``master_function``.
+
+    """
+    os.setsid()   # new session; this process is the controller
+
+    master_fd, child_fd = os.openpty()
+    pty_name = os.ttyname(child_fd)
+
+    # take controlling terminal
+    pty_fd = os.open(pty_name, os.O_RDWR)
+    os.close(pty_fd)
+
+    ready = multiprocessing.Value('i', False)
+    proc = multiprocessing.Process(
```

Comment:
I think `child_process` would be a better name for keeping track of which process is which (and for that matter `_child_process` could be called `_set_up_and_run_child_function`)

Reply:
Done.